### PR TITLE
Paint setup/surface paint

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/interface/viewport.c [hunk 1]
```diff
@@ -1537,6 +1539,1056 @@ void viewport_park_entrance_paint_setup(uint8 direction, int height, rct_map_ele
...
+			RCT2_CALLPROC_X(0x66039B, 0x99999999, 0, 0xBBBBBBBB, dx / 16, 0xDDDDDDDD, 0, 0);
+			RCT2_CALLPROC_X(0x6604F1, 0x99999999, 0, 0xBBBBBBBB, dx / 16, 0xDDDDDDDD, 0, 0);
+			RCT2_CALLPROC_X(0x65F8B9, 0x99999999, 0, 0xBBBBBBBB, dx / 16, 0xDDDDDDDD, 0, 0);
+			RCT2_CALLPROC_X(0x65FE26, 0x99999999, 0, 0xBBBBBBBB, dx / 16, 0xDDDDDDDD, 0, 0);
+		}
+	}
+
+	if (mapElement->properties.surface.ownership & 0x0F
+		&& !(RCT2_GLOBAL(0x009DEA6F, uint8) & 1)) {
+		// loc_66195F:
```
Comment (IntelOrca) [inline src/interface/viewport.c]: I think this block is drawing the park fences, perhaps extract to another function?

File: src/interface/viewport.c [hunk 2]
```diff
@@ -1854,13 +1856,13 @@ static void sub_68B3FB(int x, int y)
 			viewport_track_paint_setup(direction, height, map_element);
 			break;
 		case MAP_ELEMENT_TYPE_SCENERY:
-			RCT2_CALLPROC_X(0x6DFF47, 0, 0, direction, height, (int)map_element, 0, 0);
+			//RCT2_CALLPROC_X(0x6DFF47, 0, 0, direction, height, (int)map_element, 0, 0);
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: Remember to remove this when this is ready for merging.

File: src/interface/viewport.c [hunk 3]
```diff
@@ -1140,6 +1097,51 @@ int sub_98198C(
...
+ * @param image_id (ebx)
+ * @param x (ax)
+ * @param y (cx)
+ * @param[out] paint (ebp)
+ * @return (!CF) success
+ */
+bool sub_68818E(uint32 image_id, uint8 x, uint8 y, paint_struct ** paint)
+{
+
+	//Not a paint struct but something similar
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: I don't think my note is really correct anymore. It definitely is a paint struct.

File: src/interface/viewport.c [hunk 4]
```diff
@@ -1140,6 +1097,51 @@ int sub_98198C(
...
 /**
+ * rct2: 68818E
+ *
+ * @param image_id (ebx)
+ * @param x (ax)
+ * @param y (cx)
+ * @param[out] paint (ebp)
+ * @return (!CF) success
+ */
+bool sub_68818E(uint32 image_id, uint8 x, uint8 y, paint_struct ** paint)
```
Comment (duncanspumpkin) [inline src/interface/viewport.c]: This function attaches a paint struct to another paint struct. Perhaps call it `paint_setup_attach_paint_struct` or something better.

File: src/interface/viewport.h [hunk 5]
```diff
@@ -76,6 +76,43 @@ enum {
 	VIEWPORT_INTERACTION_MASK_BANNER = ~(1 << (VIEWPORT_INTERACTION_ITEM_BANNER - 2)), // Note the -2 for BANNER
 };
 
+typedef struct paint_struct paint_struct;
+
+struct paint_struct{
```
Comment (duncanspumpkin) [inline src/interface/viewport.h]: It looks like there are two types of paint struct. The full `0x34` ps and the mini ps aka attached ps size `0x12`. We should probably break up the struct into these two and go through the code and make it obvious which is which. It's likely that it is used interchangeably in places.

**Comment by duncanspumpkin** [comment]
Hmm annoyingly github wont let me do my normal review technique (file too big) so im going to just have to do it as multiple comments.


https://github.com/marijnvdwerf/OpenRCT2/blob/paint-setup/surface-paint/src/interface/paint_surface.c#L49-L62
Those two arrays are identical but at different rotations. I tend to use just the one array and then do a (+2 & 3) to access the other array. Sometimes it makes it easier to read. But if it doesn't keep it as is.

**Comment by duncanspumpkin** [comment]
https://github.com/marijnvdwerf/OpenRCT2/blob/paint-setup/surface-paint/src/interface/paint_surface.c#L1193

You go to the trouble of dumping this array but you forgot to use it.

**Comment by duncanspumpkin** [comment]
Line 825 use `map_element_get_direction(mapElement)`

**Comment by duncanspumpkin** [comment]
Line 826. _ebp is `terrain_type`

**Comment by duncanspumpkin** [comment]
Line 346 `viewport_surface_paint_setup_get_ebx` this function rotates the current slope so that you only need to do drawing functions for one rotation. I suggest calling it get_virtual_slope or something like that.

**Comment by duncanspumpkin** [comment]
Line 824 just use `get_current_rotation()` in the two locations `cl` is used.

**Comment by duncanspumpkin** [comment]
Line 847. I'm not sure on this but as far as i can tell the corner_heights should be used with the initialy calculated surfaceSlope and not the neighbouring tile surfaceSlope.

**Comment by duncanspumpkin** [comment]
Line 855 - 866: could do with a tidy up renaming the dx variable to height and adding 5769 to the SPR enum.

**Comment by marijnvdwerf** [comment]
@duncanspumpkin I changed that. The original code was smart and combined only the useful corners. Made the code really complicated to understand.

**Comment by duncanspumpkin** [comment]
Line 1082. VS doesn't like this for initing a struct. Use `registers regs = { 0 };` same for line 1214

**Comment by duncanspumpkin** [comment]
Line 871 - 875 should probably add a comment stating that this is for vertical tunnels

**Comment by duncanspumpkin** [comment]
Line 883 all of these flags are known.

**Comment by duncanspumpkin** [comment]
Line 884: This is grass length. The upper bits that are masked off are interim grass lengths that are not visible.

**Comment by duncanspumpkin** [comment]
Line 264 - 283 can you label these arrays.

**Comment by duncanspumpkin** [comment]
Line 948. Can you add a comment stating this is for drawing the staff patrol areas.

**Comment by duncanspumpkin** [comment]
Line 960 di is negative if in all staff view. Perhaps just & this into a bool and loose the ebx var.

Line 969 this is not surfaceShape but di & 0x7FFF aka staff list selection tab. See `WINDOW_STAFF_LIST_TAB`

**Comment by duncanspumpkin** [comment]
You can use `staff_is_patrol_area_set` for line 962

**Comment by duncanspumpkin** [comment]
Line 966 this should fall through into line 970. This bit of code draws the patrol area not set by this staff member but is set by a staff member of this type.

**Comment by duncanspumpkin** [comment]
Line 957 ebp is the colour of the patrol area. 0x38 = blue, 0x8 = grey.

**Comment by duncanspumpkin** [comment]
```	
// Draw Staff Patrol Areas
	// loc_660D02
	if (RCT2_GLOBAL(0x009DEA50, sint16) != -1) {
		sint32 staffIndex = RCT2_GLOBAL(0x009DEA50, sint16);
		bool is_staff_list = staffIndex & 0x8000;
		uint8 staffType = staffIndex & 0x7FFF;
		sint16 x = RCT2_GLOBAL(0x009DE56A, sint16), y = RCT2_GLOBAL(0x009DE56E, sint16);

		uint32 image_id = 0x20000000;
		uint8 patrolColour = 7;

		if (!is_staff_list) {
			rct_peep* staff = GET_PEEP(staffIndex);
			if (!staff_is_patrol_area_set(staffIndex, x, y)) {
				patrolColour = 1;
			}
			staffType = staff->staff_type;
		}

		x = (x & 0x1F80) >> 7;
		y = (y & 0x1F80) >> 1;
		int offset = (x | y) >> 5;
		int bitIndex = (x | y) & 0x1F;
		int ebx = (staffType + 200) * 512;
		if (RCT2_ADDRESS(RCT2_ADDRESS_STAFF_PATROL_AREAS + ebx, uint32)[offset] & (1 << bitIndex)) {
			assert(surfaceShape < countof(byte_97B444));

			image_id |= SPR_TERRAIN_SELECTION_PATROL_AREA + byte_97B444[surfaceShape];
			image_id |= patrolColour << 19;
			sub_68818E(image_id, 0, 0, NULL);
		}
	}
```

**Comment by duncanspumpkin** [comment]
Line 984 use `gPeepSpawns`

**Comment by janisozaur** [comment]
There are some `clang`-generated warnings, can you address them as well?
```
/home/janisozaur/workspace/OpenRCT2/src/interface/paint_surface.c:50:1: warning: typedef requires a name [-Wmissing-declarations]
typedef struct corner_height
^~~~~~~
/home/janisozaur/workspace/OpenRCT2/src/interface/paint_surface.c:497:7: warning: expression result unused [-Wunused-value]
        regs.al;
        ~~~~~^~
/home/janisozaur/workspace/OpenRCT2/src/interface/paint_surface.c:883:20: warning: equality comparison with extraneous parentheses
      [-Wparentheses-equality]
                                if ((zoomLevel == 0)) {
                                     ~~~~~~~~~~^~~~
/home/janisozaur/workspace/OpenRCT2/src/interface/paint_surface.c:883:20: note: remove extraneous parentheses around the comparison to silence this
      warning
                                if ((zoomLevel == 0)) {
                                    ~          ^   ~
/home/janisozaur/workspace/OpenRCT2/src/interface/paint_surface.c:883:20: note: use '=' to turn this equality comparison into an assignment
                                if ((zoomLevel == 0)) {
                                               ^~
                                               =
```

**Comment by duncanspumpkin** [comment]
Line 1006 has that been fixed? It seems to work to me.

**Comment by marijnvdwerf** [comment]
@duncanspumpkin Not sure. It might. Removed the comment for now.

**Comment by duncanspumpkin** [comment]
You've broken appveyour probably due to the typedef change. Have a look at all the other structs for the correct format that keeps everything happy

**Comment by janisozaur** [comment]
@duncanspumpkin nope, nsis *again*. I restarted the build.

**Comment by janisozaur** [comment]
AppVeyor is green now, but there are warnings:
```
src\interface\paint_surface.c(408): warning C4101: 'regs': unreferenced local variable [C:\projects\openrct2-ject9\openrct2.vcxproj] 
src\interface\paint_surface.c(1324): warning C4244: 'function': conversion from 'sint16' to 'sint8', possible loss of data [C:\projects\openrct2-ject9\openrct2.vcxproj] 
src\interface\paint_surface.c(877): warning C4101: 'regs': unreferenced local variable [C:\projects\openrct2-ject9\openrct2.vcxproj]
```

**Comment by janisozaur** [comment]
Both CIs happy now :+1:

**Comment by duncanspumpkin** [comment]
Line 1072 that should be local_surfaceShape & 0x10 (Its checking if its the really steep slope.)

**Comment by duncanspumpkin** [comment]
Line 1080 why has this been swapped? It messes up for water tools with it that way round.

**Comment by duncanspumpkin** [comment]
Line 1097-1106 change it to a for loop and loose `edi`.

**Comment by duncanspumpkin** [comment]
Line 1096 ebp not used anymore.

**Comment by duncanspumpkin** [comment]
Line 1118 technically you can only have 1 tile that will match the current tile so you can exit this loop as soon as you have finished the drawing.

**Comment by duncanspumpkin** [comment]
Line 1112 that should be !(RCT2_GLOBAL(RCT2_ADDRESS_MAP_SELECTION_FLAGS, uint8) & 8)
Line 1114 should be eax = 0x1280000; its a different colour of tile for when you are placing the first section of a ride or track.

**Comment by duncanspumpkin** [comment]
Line 1141 can you use `base_image = byte_97B84A[terrain_type];` then its a bit more obvious what its doing.

**Comment by duncanspumpkin** [comment]
Line 1157 before there was a copy of some other elements into 0x9E329C any reason why that was changed?

**Comment by duncanspumpkin** [comment]
Line 555: Can you add a rct2 address label `0x0065EB7D` and `0x0065F0D8`

**Comment by duncanspumpkin** [comment]
Line 1174-1175 can you rename them `waterHeight` and `localHeight`

**Comment by duncanspumpkin** [comment]
Line 709 and 713 can you add the rct2 address'

**Comment by duncanspumpkin** [comment]
Also why is `viewport_surface_draw_water_side_top` blank?

**Comment by marijnvdwerf** [comment]
> Line 1080 why has this been swapped? It messes up for water tools with it that way round.
It actually messes it up when I switch it.

> Also why is viewport_surface_draw_water_side_top blank?
> Line 1157 before there was a copy of some other elements into 0x9E329C any reason why that was changed?

Was attempting to do it smarter than the current buggy version. Surface edges are almost the same function though, so just check those and I'll use the original.

**Comment by marijnvdwerf** [comment]
Water selection broke yesterday. 
image

**Comment by duncanspumpkin** [comment]
@marijnvdwerf you still have a mistake on line 1268 that should be `& 0x10` instead of `== 0x10`

**Comment by duncanspumpkin** [comment]
Line 1401 can you add a little description stating that this does the park boundaries fences.

**Comment by duncanspumpkin** [comment]
Line 1520 you have missed a statement where the global var RCT2_ADDRESS_PAINT_SETUP_CURRENT_TYPE is switched back to terrain. I realise its not of much use at this point though as there is no more drawing but can you keep it in for now.

**Comment by duncanspumpkin** [comment]
Line 994: Youve remove a copying of the mapElement pointer into 0x9E3250 this is used by other drawing functions and will cause a crash on certain track paint functions

**Comment by duncanspumpkin** [comment]
Line 526 `>=`

**Comment by duncanspumpkin** [comment]
Line 1392 - 1395 that should be water height not height

**Comment by marijnvdwerf** [comment]
526 is never equal. So that was a conscious one.

**Comment by marijnvdwerf** [comment]
image

**Comment by marijnvdwerf** [comment]
@duncanspumpkin Done, I think.

**Comment by janisozaur** [comment]
nice!

**Pull request merged**